doc/sonic-vrf-hld-v0.2.md

```diff
@@ -0,0 +1,412 @@
+
+;Stores a list of routes
+
+;Status: Mandatory
+
+key = ROUTE_TABLE:VRF_NAME:prefix ;
+
+nexthop = \*prefix, ;IP addresses separated “,” (empty indicates no gateway)
+
+intf = ifindex? PORT_TABLE.key ; zero or more separated by “,” (zero indicates
+no interface)
+
+blackhole = BIT ; Set to 1 if this route is a blackhole (or null0)
+
+Since global vrf name is null, global vrf key will becomes ROUTE_TABLE:prefix.
+
+ VrfMgrd 
+---------
+
+>   Listening VRF related configuration in config_db such as VRF
+>   creation/deletion, VRF binding to any interface. Once detected, update
+>   kernel using iproute2 CLIs and write VRF information to app-VRF-table and
+>   app-intf-table.
+
+>   VrfMgrd process will be placed in swss docker. In case of “swss restart”,
+>   VRF device will be still retained in kernel. When VrfMgrd starts up it
+>   querys all master device from kernel and clean up all vrf related device and
+>   restore vrf device per configdb.
+
+>   VRFOrch has three member routerCnt , neighCnt and rifCnt which record VRF
+>   related route number , neigh number and rif number.VRFOrch is added to
+>   RouteOrch , NeighOrch and intfsOrch member to update routeCnt , neighCnt and
+>   rifCnt.
+
+>   When  VRFOrch receives vrf-delete event VRF object won’t  be deleted until
+>   routerCnt  ,neighCnt and rifCnt  is decreased to zero.
+
+>   When device binds to specified VRF, the ip address of slave device will be
+>   removed and kernel will delete all neigh associated slave device.
+
+ fpmsyncd changes
+-----------------
+
+with added VRF ID, fpmsyncd can use rtnl_route_get_table to acquire tableid.
+Hence can send VRF routes further down. The messages from FRR has nh(next hop)
+information which contain further information about (nexthop_ipaddress and
+interface index)，tableid can be derived from the interface index.
+
+Fpmsyncd can build \<tableid, vrf_name\> pairs using rtnetlink api.
+
+ adding vrforch 
+----------------
+
+>   Monitorying app-VRF-table，Using sai_create_virtual_router_fn or
+
+>   sai_remove_virtual_router_fn defined in saivirtualrouter.h to track
+>   (VR，VRF) creation/deletion.and save (vrf_name, vrf-vid) pairs.
+
+intfsorch changes
+-----------------
+
+>   Adding following logics:
+
+-   adding vrforch as one member to intfsorch
+
+-   intfsorch monitors both app-intf-table和app-intf-prefix-table，when
+    app-intf-table has changes，handle updating vrf attribute on
+    routerintf，request vrforch for tableid/vrf-id .
+
+    1.  routeorch changes
+
+Adding the following logics:
+
+-   Adding vrforch member to routesorch
+
+-   Once app-route-table has new udpate，get tableid from vrforch for route
+    add/delete.
+
+>   When query nexthop，keys now are (tableid, ipaddress)，tableid of nexthop
+>   can be acquired using nexthop interface.
+
+neighorch changes
+-----------------
+
+Adding the following logic:
+
+-   the Key of NextHop now is changed from only ipaddress to a pair of
+    (ipaddress, interface_name)
```

Comment:
Should be (ipaddress, tableid) and not (ipaddress, interface_name). Expect to be doing a lot of conversions and lookups if you leave as (ipaddress, interface_name).

Reply:
can we point a route in vrf red to a nexthop (neighbor) in vrf blue? @prsunny , do we need to change the nexthop format in app db?

Reply:
@lguohan Yes you can have routes in vrf red with nexthops in vrf blue.

Reply:
@nikos-github , yes it is possible but the format in APP_DB is same for route since every route has a nexthop IP with the interface name. Currently orchagent doesn't look for the interface which requires a fix.

Reply:
@nikos-github When set neigh to SAI layer (ipaddress, interface_name) would be more comfortable. And relatedAPP_DB/ASIC_DB needn't be changed.